go/internal/feast/onlinestore/cassandraonlinestore.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to the ScyllaDB database")
+	}
+	store.session = createdSession
+	return &store, nil
+}
+
+func (c *CassandraOnlineStore) getFqTableName(tableName string) string {
+	return fmt.Sprintf(`"%s"."%s_%s"`, c.keyspace, c.project, tableName)
+}
+
+func (c *CassandraOnlineStore) getCQLStatement(tableName string, featureNames []string, nKeys int) string {
+	// TODO: Compare with multiple single-key concurrent queries like in the Python feature server
+	keyPlaceholders := make([]string, nKeys)
+	for i := 0; i < nKeys; i++ {
+		keyPlaceholders[i] = "?"
+	}
+
+	// this prevents fetching unnecessary features
+	quotedFeatureNames := make([]string, len(featureNames))
+	for i, featureName := range featureNames {
+		quotedFeatureNames[i] = fmt.Sprintf(`'%s'`, featureName)
+	}
+
+	return fmt.Sprintf(
+		`SELECT "entity_key", "feature_name", "event_ts", "value" FROM %s WHERE "entity_key" IN (%s) AND "feature_name" IN (%s)`,
+		tableName,
+		strings.Join(keyPlaceholders, ","),
+		strings.Join(quotedFeatureNames, ","),
+	)
+}
+
+func (c *CassandraOnlineStore) buildCassandraEntityKeys(entityKeys []*types.EntityKey) ([]interface{}, map[string]int, error) {
+	cassandraKeys := make([]interface{}, len(entityKeys))
+	cassandraKeyToEntityIndex := make(map[string]int)
+	for i := 0; i < len(entityKeys); i++ {
+		var key, err = utils.SerializeEntityKey(entityKeys[i], c.config.EntityKeySerializationVersion)
+		if err != nil {
+			return nil, nil, err
+		}
+		// encoding to hex
+		encodedKey := hex.EncodeToString(*key)
+		cassandraKeys[i] = encodedKey
+		cassandraKeyToEntityIndex[encodedKey] = i
+	}
+	return cassandraKeys, cassandraKeyToEntityIndex, nil
+}
+func (c *CassandraOnlineStore) OnlineRead(ctx context.Context, entityKeys []*types.EntityKey, featureViewNames []string, featureNames []string) ([][]FeatureData, error) {
+	uniqueNames := make(map[string]int32)
+	for _, fvName := range featureViewNames {
+		uniqueNames[fvName] = 0
+	}
+	if len(uniqueNames) != 1 {
+		return nil, fmt.Errorf("rejecting OnlineRead as more than 1 feature view was tried to be read at once")
+	}
+
+	serializedEntityKeys, serializedEntityKeyToIndex, err := c.buildCassandraEntityKeys(entityKeys)
+
+	if err != nil {
+		return nil, fmt.Errorf("error when serializing entity keys for Cassandra")
+	}
+	results := make([][]FeatureData, len(entityKeys))
+	for i := range results {
+		results[i] = make([]FeatureData, len(featureNames))
+	}
+
+	featureNamesToIdx := make(map[string]int)
+	for idx, name := range featureNames {
+		featureNamesToIdx[name] = idx
+	}
+
+	featureViewName := featureViewNames[0]
+
+	// Prepare the query
+	tableName := c.getFqTableName(featureViewName)
+	cqlStatement := c.getCQLStatement(tableName, featureNames, len(serializedEntityKeys))
+	// Bundle the entity keys in one statement (gocql handles this as concurrent queries)
+	scanner := c.session.Query(cqlStatement, serializedEntityKeys...).Iter().Scanner()
```

Comment:
We will easily be impacted the default page size set by client. https://github.com/scylladb/gocql/blob/30a838e39b1f0e4d559989b9f30404c5399b3d5e/cluster.go#L142-L144C2
We have a clients requesting 200-300 features and 100 keys (if they do batching at orchestrator. If not we get more keys.

Reply:
Even if we partition queries, at one point the results of all those smaller queries will be in memory. I think such clients will have to request larger pods when deploying their feature servers.